### PR TITLE
Handle FileNotFoundException is thrown when attempting to send a file

### PR DESCRIPTION
**Pull request by natemcmaster**
Resolves aspnet/Home#2775

If a file disappears between the File.Exists and File.Open, or if the file is a broken symlink, static files throws an uncaught exception, causing HTTP 500. This change catches the FileNotFoundException and continues up the middleware stack.

It also handles the case when an implementation of IHttpSendFileFeature may throw FileNotFoundException.

### CODE REVIEW

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs [hunk 1]
```diff
@@ -97,21 +98,35 @@ public Task Invoke(HttpContext context)
...
                         }
                         if (fileContext.IsRangeRequest)
                         {
-                            return fileContext.SendRangeAsync();
+                            await fileContext.SendRangeAsync();
+                            return;
                         }
-                        _logger.LogFileServed(fileContext.SubPath, fileContext.PhysicalPath);
-                        return fileContext.SendAsync();
+                        try
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs]: Expand the try to include SendRangeAsync

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs [hunk 2]
```diff
@@ -97,21 +98,35 @@ public Task Invoke(HttpContext context)
...
+                        {
+                            await fileContext.SendAsync();
+                            _logger.LogFileServed(fileContext.SubPath, fileContext.PhysicalPath);
+                            return;
+                        }
+                        catch (FileNotFoundException)
+                        {
+                            context.Response.Clear();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs]: Comment: Fall through to the next middleware as if the file never existed.

File: test/Microsoft.AspNetCore.StaticFiles.Tests/StaticFileMiddlewareTests.cs [hunk 3]
```diff
@@ -29,6 +33,52 @@ public class StaticFileMiddlewareTests
...
+            Assert.Null(response.Headers.ETag);
+        }
+
+        [ConditionalFact]
+        [OSSkipCondition(OperatingSystems.Windows, SkipReason = "Symlinks not supported on Windows")]
+        public async Task ReturnsNotFoundForBrokenSymlink()
+        {
+            var badLink = Path.Combine(AppContext.BaseDirectory, Path.GetRandomFileName() + ".txt");
+
+            Process.Start("ln", $"-s \"/tmp/{Path.GetRandomFileName()}\" \"{badLink}\"").WaitForExit();
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.StaticFiles.Tests/StaticFileMiddlewareTests.cs]: try/finally delete the link?

Comment (natemcmaster) [reply test/Microsoft.AspNetCore.StaticFiles.Tests/StaticFileMiddlewareTests.cs]: Updated.

**Pull request merged**